File: backend/packages/users/src/utils/clients.py
from typing import Any
import psycopg2
import psycopg2.extras
import redis


class Singleton(type):
    _instances = {}

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        if self not in self._instances:
            self._instances[self] = super().__call__(*args, **kwargs)

        return self._instances[self]


class Clients:
    def __init__(
        self,
        redis_host: str,
        redis_port: str,
        pg_host: str,
        pg_port: str,
        pg_db: str,
        pg_user: str,
        pg_password: str,
    ) -> None:
        self._redis_client = redis.Redis(host=redis_host, port=redis_port)
        self._pg_conn = psycopg2.connect(
            host=pg_host,
            port=pg_port,
            database=pg_db,
            user=pg_user,
            password=pg_password,
        )

    @property
    def cache(self):
        return self._redis_client

    @property
    def database(self):
        return self._pg_conn


class ClientsSingleton(Clients, metaclass=Singleton):
    pass

File: backend/packages/projects/src/utils/clients.ts
import redis from 'redis'
import pg from 'pg'
import aws from 'aws-sdk'

type Clients = {
    callbackAPIClient: aws.ApiGatewayManagementApi
    redisClient: redis.RedisClientType
    postgresClient: pg.Pool
}
type Singleton<T> = { current: null | T }

const instance: Singleton<Clients> = { current: null }

export const makeClients = async (
    callbackApiDomainName?: string,
    callbackApiStage?: string,
    redisHost?: string,
    redisPort?: number,
    databaseHost?: string,
    databasePort?: number,
    databaseDatabase?: string,
    databaseUsername?: string,
    databasePassword?: string,
) => {
    if (!instance.current) {
        const callbackAPIClient = new aws.ApiGatewayManagementApi({
            apiVersion: '2018-11-29',
            endpoint: callbackApiDomainName + '/' + callbackApiStage,
        })
        const redisClient = await redis.createClient({ url: `redis://${redisHost}:${redisPort}` }).connect() as any
        const postgresClient = new pg.Pool({
            host: databaseHost,
            port: databasePort,
            database: databaseDatabase,
            user: databaseUsername,
            password: databasePassword
        })

        instance.current = {
            callbackAPIClient,
            redisClient,
            postgresClient
        }
    }

    return instance.current
}
